src/uu/tail/src/tail.rs

```diff
@@ -645,14 +1413,131 @@ fn parse_num(src: &str) -> Result<(u64, bool), ParseSizeError> {
     parse_size(size_string).map(|n| (n, starting_with))
 }
 
-fn get_block_size(md: &Metadata) -> u64 {
+pub fn stdin_is_pipe_or_fifo() -> bool {
     #[cfg(unix)]
     {
-        md.blocks()
+        platform::stdin_is_pipe_or_fifo()
     }
-    #[cfg(not(unix))]
+    #[cfg(windows)]
+    {
+        winapi_util::file::typ(winapi_util::HandleRef::stdin())
+            .map(|t| t.is_disk() || t.is_pipe())
+            .unwrap_or(false)
+    }
+}
+
+pub fn stdin_is_bad_fd() -> bool {
+    #[cfg(unix)]
     {
-        md.len()
+        platform::stdin_is_bad_fd()
+    }
+    #[cfg(not(unix))]
+    false
+}
+
+trait FileExtTail {
```

Comment:
All these additional `trait`s are excellent and really help wit readability in the rest of the code. Well done!